python/lsst/ip/isr/deferredCharge.py

```diff
@@ -0,0 +1,639 @@
+                                                           self.config.nPixelTrapCorrection)
+
+                # Undo flips here.  The method is symmetric.
+                correctedAmpData = self.flipData(correctedAmpData, amp)
+                image[amp.getBBox()].getArray()[:, :] = correctedAmpData[:, :]
+
+        return exposure
+
+    @staticmethod
+    def flipData(ampData, amp):
+        """Flip data array such that readout corner is at lower-left.
+
+        Parameters
+        ----------
+        ampData : `np.ndarray`, (nx, ny)
+            Image data to flip.
+        amp : `lsst.afw.cameraGeom.Amplifier`
+            Amplifier to get readout corner information.
+
+        Returns
+        -------
+        ampData : `np.ndarray`, (nx, ny)
+            Flipped image data.
+        """
+        X_FLIP = {ReadoutCorner.LL: False,
+                  ReadoutCorner.LR: True,
+                  ReadoutCorner.UL: False,
+                  ReadoutCorner.UR: True}
+        Y_FLIP = {ReadoutCorner.LL: False,
+                  ReadoutCorner.LR: False,
+                  ReadoutCorner.UL: True,
+                  ReadoutCorner.UR: True}
+
+        if X_FLIP(amp.getReadoutCorner()):
+            ampData = np.fliplr(ampData)
+        if Y_FLIP(amp.getReadoutCorner()):
+            ampData = np.flipud(ampData)
+
+        return ampData
+
+    @staticmethod
+    def local_offset_inverse(inputArr, drift_scale, decay_time, num_previous_pixels=15):
+        """Remove CTI effects from local offsets.
+
+        This implements equation 10 of Snyder+21.  For an image with
+        CTI, s'(m, n), the correction factor is equal to the maximum
+        value of the set of:
+            {A_L s'(m, n - j) exp(-j t / \tau_L)}_j=0^jmax
+
+        Parameters
+        ----------
+        inputArr : `np.ndarray`, (nx, ny)
+            Input image data to correct.
+        drift_scale : `float`
+            Drift scale (Snyder+21 A_L value) to use in correction.
+        decay_time : `float`
+            Decay time (Snyder+21 \tau_L) of the correction.
+        num_previous_pixels : `int`, optional
+            Number of previous pixels to use for correction.  As the
+            CTI has an exponential decay, this essentially truncates
+            the correction where that decay scales the input charge to
+            near zero.
+
+        Returns
+        -------
+        outputArr : `np.ndarray`, (nx, ny)
+            Corrected image data.
+        """
+        r = np.exp(-1/decay_time)
+        Ny, Nx = inputArr.shape
+
+        # j = 0 term:
+        offset = np.zeros((num_previous_pixels, Ny, Nx))
+        offset[0, :, :] = drift_scale*np.maximum(0, inputArr)
+
+        # j = 1..jmax terms:
+        for n in range(1, num_previous_pixels):
+            offset[n, :, n:] = drift_scale*np.maximum(0, inputArr[:, :-n])*(r**n)
+
+        Linv = np.amax(offset, axis=0)
+        outputArr = inputArr - Linv
```

Comment:
So if I'm reading this right, this locates a whole row, the one containing the highest value in the 2d array, and then subtracts that whole row off each row in the original 2d array. Is that right?